src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+                }
+
+                public override Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync, CancellationToken cancellationToken)
+                {
+                    return _connection.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync, cancellationToken);
+                }
+
+                protected override void OnDisposed()
+                {
+                    _pools.Free(_serviceName, _connection);
+                }
+            }
+        }
+
+        private static class Connections
+        {
+            /// <summary>
+            /// call <paramref name="funcAsync"/> and retry up to <paramref name="timeout"/> if the call throws
+            /// <typeparamref name="TException"/>. any other exception from the call won't be handled here.
+            /// </summary>
+            public static async Task<TResult> RetryRemoteCallAsync<TException, TResult>(
+                Func<Task<TResult>> funcAsync,
+                TimeSpan timeout,
+                CancellationToken cancellationToken) where TException : Exception
+            {
+                const int retry_delayInMS = 50;
+
+                using (var pooledStopwatch = SharedPools.Default<Stopwatch>().GetPooledObject())
+                {
+                    var watch = pooledStopwatch.Object;
+                    watch.Start();
+
+                    while (watch.Elapsed < timeout)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        try
+                        {
+                            return await funcAsync().ConfigureAwait(false);
+                        }
+                        catch (TException)
+                        {
+                            // throw cancellation token if operation is cancelled
+                            cancellationToken.ThrowIfCancellationRequested();
+                        }
+
+                        // wait for retry_delayInMS before next try
+                        await Task.Delay(retry_delayInMS, cancellationToken).ConfigureAwait(false);
+
+                        ReportTimeout(watch);
+                    }
+                }
+
+                // operation timed out, more than we are willing to wait
+                ShowInfoBar();
+
+                // user didn't ask for cancellation, but we can't fullfill this request. so we
+                // create our own cancellation token and then throw it. this doesn't guarantee
+                // 100% that we won't crash, but this is at least safest way we know until user
+                // restart VS (with info bar)
+                using (var ownCancellationSource = new CancellationTokenSource())
+                {
+                    ownCancellationSource.Cancel();
+                    ownCancellationSource.Token.ThrowIfCancellationRequested();
```

Comment:
this seems *really* janky.  Won't this just cause aggregate exceptions somewhere?  Why not call this TryRetryRemoteCallAsync, and return 'default' in the event that we get here?

Reply:
everything thing under Connections are just moved from ServiceHubRemoteHostClient. I am not going to touch anything there in this PR.

...

sure this could. but this is the way we choose to do soft crash. we are not interested on recovering. we already showed info bar to shutdown VS. we are only interested in not crash VS.

Reply:
> we are only interested in not crash VS.

I'm a bit confused.  As this can become an AggregateException, it seems *more* likely that a crash would happen after this.

Reply:
if we find such case, we will make that part not to crash due to aggregate exception.
..

no. this has been here for months since the beginning of OOP. and we didn't find VS crashes due to this. even if we see such case, our fix will be changing that code not to crash by dealing with AggregateException.